// @probat/react/src/components/Experiment.tsx
"use client";

import React, {
    useEffect,
    useRef,
    useState,
    useCallback,
    useMemo,
} from "react";
import { useProbatContext } from "../context/ProbatContext";
import { fetchDecision, sendMetric, extractClickMeta } from "../utils/api";
import { getDistinctId, getPageKey } from "../utils/eventContext";
import { makeDedupeKey, hasSeen, markSeen } from "../utils/dedupeStorage";
import { useStableInstanceId } from "../utils/stableInstanceId";

// ── localStorage assignment cache ──────────────────────────────────────────

const ASSIGNMENT_PREFIX = "probat:assignment:";

interface StoredAssignment {
    variantKey: string;
    ts: number;
}

function readAssignment(id: string): string | null {
    if (typeof window === "undefined") return null;
    try {
        const raw = localStorage.getItem(ASSIGNMENT_PREFIX + id);
        if (!raw) return null;
        const parsed: StoredAssignment = JSON.parse(raw);
        return parsed.variantKey ?? null;
    } catch {
        return null;
    }
}

function writeAssignment(id: string, variantKey: string): void {
    if (typeof window === "undefined") return;
    try {
        const entry: StoredAssignment = { variantKey, ts: Date.now() };
        localStorage.setItem(ASSIGNMENT_PREFIX + id, JSON.stringify(entry));
    } catch {}
}

// ── Types ──────────────────────────────────────────────────────────────────

export interface ExperimentTrackOptions {
    /** Auto-track impressions (default true) */
    impression?: boolean;
    /** Auto-track clicks (default true) */
    primaryClick?: boolean;
    /** Custom impression event name (default "$experiment_exposure") */
    impressionEventName?: string;
    /** Custom click event name (default "$experiment_click") */
    clickEventName?: string;
}

export interface ExperimentProps {
    /** Experiment key / identifier */
    id: string;
    /** Control variant ReactNode */
    control: React.ReactNode;
    /** Named variant ReactNodes, keyed by variant key (e.g. { ai_v1: <MyVariant /> }) */
    variants: Record<string, React.ReactNode>;
    /** Tracking configuration */
    track?: ExperimentTrackOptions;
    /** Stable instance id when multiple instances of the same experiment exist on a page */
    componentInstanceId?: string;
    /** Behavior when assignment fetch fails: "control" (default) renders control, "suspend" throws */
    fallback?: "control" | "suspend";
    /** Log decisions + events to console */
    debug?: boolean;
}

export function Experiment({
    id,
    control,
    variants,
    track,
    componentInstanceId,
    fallback = "control",
    debug = false,
}: ExperimentProps) {
    const { host, bootstrap } = useProbatContext();

    // Stable instance id (useId + sessionStorage for cross-mount stability)
    const autoInstanceId = useStableInstanceId(id);
    const instanceId = componentInstanceId ?? autoInstanceId;

    // Track options with defaults
    const trackImpression = track?.impression !== false;
    const trackClick = track?.primaryClick !== false;
    const impressionEvent = track?.impressionEventName ?? "$experiment_exposure";
    const clickEvent = track?.clickEventName ?? "$experiment_click";

    // ── Assignment resolution ──────────────────────────────────────────────

    const [variantKey, setVariantKey] = useState<string>(() => {
        // Sync resolution order: bootstrap → localStorage → "control"
        if (bootstrap[id]) return bootstrap[id];
        const cached = readAssignment(id);
        if (cached) return cached;
        return "control";
    });
    const [resolved, setResolved] = useState<boolean>(() => {
        return !!(bootstrap[id] || readAssignment(id));
    });

    useEffect(() => {
        // Already resolved from bootstrap or cache
        if (bootstrap[id] || readAssignment(id)) {
            // Ensure state is synced (StrictMode may re-mount)
            const key = bootstrap[id] ?? readAssignment(id) ?? "control";
            setVariantKey(key);
            setResolved(true);
            return;
        }

        let cancelled = false;

        (async () => {
            try {
                const distinctId = getDistinctId();
                const key = await fetchDecision(host, id, distinctId);
                if (cancelled) return;

                // Validate variant key
                if (key !== "control" && !(key in variants)) {
                    if (debug) {
                        console.warn(
                            `[probat] Unknown variant "${key}" for experiment "${id}", falling back to control`
                        );
                    }
                    setVariantKey("control");
                } else {
                    setVariantKey(key);
                    writeAssignment(id, key);
                }
            } catch (err) {
                if (cancelled) return;
                if (debug) {
                    console.error(`[probat] fetchDecision failed for "${id}":`, err);
                }
                if (fallback === "suspend") throw err;
                setVariantKey("control");
            } finally {
                if (!cancelled) setResolved(true);
            }
        })();

        return () => {
            cancelled = true;
        };
    }, [id, host]); // eslint-disable-line react-hooks/exhaustive-deps

    // ── Debug logging ──────────────────────────────────────────────────────

    useEffect(() => {
        if (debug && resolved) {
            console.log(`[probat] Experiment "${id}" → variant "${variantKey}"`, {
                instanceId,
                pageKey: getPageKey(),
            });
        }
    }, [debug, id, variantKey, resolved, instanceId]);

    // ── Shared event properties ────────────────────────────────────────────

    const eventProps = useMemo(
        () => ({
            experiment_id: id,
            variant_key: variantKey,
            component_instance_id: instanceId,
        }),
        [id, variantKey, instanceId]
    );

    // ── Impression tracking via IntersectionObserver ────────────────────────

    const containerRef = useRef<HTMLDivElement>(null);
    const impressionSent = useRef(false);

    useEffect(() => {
        if (!trackImpression || !resolved) return;

        // Reset on re-mount (StrictMode safety)
        impressionSent.current = false;

        const pageKey = getPageKey();
        const dedupeKey = makeDedupeKey(id, variantKey, instanceId, pageKey);

        // Already seen this session
        if (hasSeen(dedupeKey)) {
            impressionSent.current = true;
            return;
        }

        const el = containerRef.current;
        if (!el) return;

        // Fallback: no IntersectionObserver (SSR, old browser)
        if (typeof IntersectionObserver === "undefined") {
            if (!impressionSent.current) {
                impressionSent.current = true;
                markSeen(dedupeKey);
                sendMetric(host, impressionEvent, eventProps);
                if (debug) console.log(`[probat] Impression sent (no IO) for "${id}"`);
            }
            return;
        }

        let timer: ReturnType<typeof setTimeout> | null = null;

        const observer = new IntersectionObserver(
            ([entry]) => {
                if (!entry || impressionSent.current) return;

                if (entry.isIntersecting) {
                    timer = setTimeout(() => {
                        if (impressionSent.current) return;
                        impressionSent.current = true;
                        markSeen(dedupeKey);
                        sendMetric(host, impressionEvent, eventProps);
                        if (debug) console.log(`[probat] Impression sent for "${id}"`);
                        observer.disconnect();
                    }, 250);
                } else if (timer) {
                    clearTimeout(timer);
                    timer = null;
                }
            },
            { threshold: 0.5 }
        );

        observer.observe(el);

        return () => {
            observer.disconnect();
            if (timer) clearTimeout(timer);
        };
    }, [
        trackImpression,
        resolved,
        id,
        variantKey,
        instanceId,
        host,
        impressionEvent,
        eventProps,
        debug,
    ]);

    // ── Click tracking ─────────────────────────────────────────────────────

    const handleClick = useCallback(
        (e: React.MouseEvent) => {
            if (!trackClick) return;

            const meta = extractClickMeta(e.target as EventTarget);
            if (!meta) return;

            sendMetric(host, clickEvent, {
                ...eventProps,
                ...meta,
            });
            if (debug) {
                console.log(`[probat] Click tracked for "${id}"`, meta);
            }
        },
        [trackClick, host, clickEvent, eventProps, id, debug]
    );

    // ── Render ─────────────────────────────────────────────────────────────

    const content =
        variantKey === "control" || !(variantKey in variants)
            ? control
            : variants[variantKey];

    return (
        <div
            ref={containerRef}
            onClick={handleClick}
            data-probat-experiment={id}
            data-probat-variant={variantKey}
            style={{
                display: "block",
                margin: 0,
                padding: 0,
                opacity: resolved ? 1 : 0,
                transition: resolved ? "opacity 0.15s ease-in" : "none",
            }}
        >
            {content}
        </div>
    );
}
