// @probat/react/src/__tests__/setup.ts
import "@testing-library/jest-dom/vitest";

// Mock IntersectionObserver
class MockIntersectionObserver implements IntersectionObserver {
    readonly root: Element | null = null;
    readonly rootMargin: string = "";
    readonly thresholds: ReadonlyArray<number> = [];

    private callback: IntersectionObserverCallback;
    private elements: Set<Element> = new Set();

    constructor(callback: IntersectionObserverCallback, _options?: IntersectionObserverInit) {
        this.callback = callback;
        MockIntersectionObserver._instances.push(this);
    }

    observe(el: Element) {
        this.elements.add(el);
    }

    unobserve(el: Element) {
        this.elements.delete(el);
    }

    disconnect() {
        this.elements.clear();
    }

    takeRecords(): IntersectionObserverEntry[] {
        return [];
    }

    // Test helpers
    static _instances: MockIntersectionObserver[] = [];

    static _reset() {
        MockIntersectionObserver._instances = [];
    }

    _trigger(isIntersecting: boolean) {
        const entries: IntersectionObserverEntry[] = Array.from(this.elements).map(
            (el) =>
                ({
                    target: el,
                    isIntersecting,
                    intersectionRatio: isIntersecting ? 0.6 : 0,
                    boundingClientRect: {} as DOMRectReadOnly,
                    intersectionRect: {} as DOMRectReadOnly,
                    rootBounds: null,
                    time: Date.now(),
                }) as IntersectionObserverEntry
        );
        this.callback(entries, this);
    }
}

Object.defineProperty(globalThis, "IntersectionObserver", {
    value: MockIntersectionObserver,
    writable: true,
});

// Export for test use
export { MockIntersectionObserver };
