// @probat/react/src/context/ProbatContext.tsx
"use client";

import React, { createContext, useContext, useMemo } from "react";

export interface ProbatContextValue {
    host: string;
    userId: string;
    bootstrap: Record<string, string>;
}

const ProbatContext = createContext<ProbatContextValue | null>(null);

const DEFAULT_HOST = "https://gushi.onrender.com";

export interface ProbatProviderProps {
    /** Gushi user ID (UUID) — used to scope SDK requests to a customer */
    userId: string;
    /** Base URL for the Probat API. Defaults to https://gushi.onrender.com */
    host?: string;
    /**
     * Bootstrap assignments to avoid flash on first render.
     * Map of experiment id → variant key.
     * e.g. { "cta-copy-test": "ai_v1" }
     */
    bootstrap?: Record<string, string>;
    children: React.ReactNode;
}

export function ProbatProvider({
    userId,
    host = DEFAULT_HOST,
    bootstrap,
    children,
}: ProbatProviderProps) {
    const value = useMemo<ProbatContextValue>(
        () => ({
            host: host.replace(/\/$/, ""),
            userId,
            bootstrap: bootstrap ?? {},
        }),
        [userId, host, bootstrap]
    );

    return (
        <ProbatContext.Provider value={value}>
            {children}
        </ProbatContext.Provider>
    );
}

export function useProbatContext(): ProbatContextValue {
    const ctx = useContext(ProbatContext);
    if (!ctx) {
        throw new Error(
            "useProbatContext must be used within <ProbatProviderClient>. Wrap your app with <ProbatProviderClient userId={...}>."
        );
    }
    return ctx;
}
