// @probat/react/dist/index.js
"use client";
"use client";
'use strict';

var React3 = require('react');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React3__default = /*#__PURE__*/_interopDefault(React3);

var ProbatContext = React3.createContext(null);
var DEFAULT_HOST = "https://gushi.onrender.com";
function ProbatProvider({
  userId,
  host = DEFAULT_HOST,
  bootstrap,
  children
}) {
  const value = React3.useMemo(
    () => ({
      host: host.replace(/\/$/, ""),
      userId,
      bootstrap: bootstrap ?? {}
    }),
    [userId, host, bootstrap]
  );
  return /* @__PURE__ */ React3__default.default.createElement(ProbatContext.Provider, { value }, children);
}
function useProbatContext() {
  const ctx = React3.useContext(ProbatContext);
  if (!ctx) {
    throw new Error(
      "useProbatContext must be used within <ProbatProviderClient>. Wrap your app with <ProbatProviderClient userId={...}>."
    );
  }
  return ctx;
}

// src/components/ProbatProviderClient.tsx
function ProbatProviderClient(props) {
  return React3__default.default.createElement(ProbatProvider, props);
}

// src/utils/environment.ts
function detectEnvironment() {
  if (typeof window === "undefined") {
    return "prod";
  }
  const hostname = window.location.hostname;
  if (hostname === "localhost" || hostname === "127.0.0.1" || hostname === "0.0.0.0" || hostname.startsWith("192.168.") || hostname.startsWith("10.") || hostname.startsWith("172.16.") || hostname.startsWith("172.17.") || hostname.startsWith("172.18.") || hostname.startsWith("172.19.") || hostname.startsWith("172.20.") || hostname.startsWith("172.21.") || hostname.startsWith("172.22.") || hostname.startsWith("172.23.") || hostname.startsWith("172.24.") || hostname.startsWith("172.25.") || hostname.startsWith("172.26.") || hostname.startsWith("172.27.") || hostname.startsWith("172.28.") || hostname.startsWith("172.29.") || hostname.startsWith("172.30.") || hostname.startsWith("172.31.")) {
    return "dev";
  }
  return "prod";
}

// src/utils/eventContext.ts
var DISTINCT_ID_KEY = "probat:distinct_id";
var SESSION_ID_KEY = "probat:session_id";
var cachedDistinctId = null;
var cachedSessionId = null;
function generateId() {
  if (typeof crypto !== "undefined" && crypto.randomUUID) {
    return crypto.randomUUID();
  }
  const bytes = new Uint8Array(16);
  if (typeof crypto !== "undefined" && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
  }
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}
function getDistinctId() {
  if (cachedDistinctId) return cachedDistinctId;
  if (typeof window === "undefined") return "server";
  try {
    const stored = localStorage.getItem(DISTINCT_ID_KEY);
    if (stored) {
      cachedDistinctId = stored;
      return stored;
    }
  } catch {
  }
  const id = `anon_${generateId()}`;
  cachedDistinctId = id;
  try {
    localStorage.setItem(DISTINCT_ID_KEY, id);
  } catch {
  }
  return id;
}
function getSessionId() {
  if (cachedSessionId) return cachedSessionId;
  if (typeof window === "undefined") return "server";
  try {
    const stored = sessionStorage.getItem(SESSION_ID_KEY);
    if (stored) {
      cachedSessionId = stored;
      return stored;
    }
  } catch {
  }
  const id = `sess_${generateId()}`;
  cachedSessionId = id;
  try {
    sessionStorage.setItem(SESSION_ID_KEY, id);
  } catch {
  }
  return id;
}
function getPageKey() {
  if (typeof window === "undefined") return "";
  return window.location.pathname + window.location.search;
}
function getPageUrl() {
  if (typeof window === "undefined") return "";
  return window.location.href;
}
function getReferrer() {
  if (typeof document === "undefined") return "";
  return document.referrer;
}
function buildEventContext() {
  return {
    distinct_id: getDistinctId(),
    session_id: getSessionId(),
    $page_url: getPageUrl(),
    $pathname: typeof window !== "undefined" ? window.location.pathname : "",
    $referrer: getReferrer()
  };
}

// src/utils/api.ts
var pendingDecisions = /* @__PURE__ */ new Map();
async function fetchDecision(host, experimentId, distinctId) {
  const existing = pendingDecisions.get(experimentId);
  if (existing) return existing;
  const promise = (async () => {
    try {
      const url = `${host.replace(/\/$/, "")}/experiment/decide`;
      const res = await fetch(url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Accept: "application/json"
        },
        credentials: "include",
        body: JSON.stringify({
          experiment_id: experimentId,
          distinct_id: distinctId
        })
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const data = await res.json();
      return data.variant_key || "control";
    } finally {
      pendingDecisions.delete(experimentId);
    }
  })();
  pendingDecisions.set(experimentId, promise);
  return promise;
}
function sendMetric(host, event, properties) {
  if (typeof window === "undefined") return;
  const ctx = buildEventContext();
  const payload = {
    event,
    environment: detectEnvironment(),
    properties: {
      ...ctx,
      source: "react-sdk",
      captured_at: (/* @__PURE__ */ new Date()).toISOString(),
      ...properties
    }
  };
  try {
    const url = `${host.replace(/\/$/, "")}/experiment/metrics`;
    fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      credentials: "include",
      body: JSON.stringify(payload)
    }).catch(() => {
    });
  } catch {
  }
}
function extractClickMeta(target) {
  if (!target || !(target instanceof HTMLElement)) return null;
  const primary = target.closest('[data-probat-click="primary"]');
  if (primary) return buildMeta(primary, true);
  const interactive = target.closest('button, a, [role="button"]');
  if (interactive) return buildMeta(interactive, false);
  return null;
}
function buildMeta(el, isPrimary) {
  const meta = {
    click_target_tag: el.tagName,
    click_is_primary: isPrimary
  };
  if (el.id) meta.click_target_id = el.id;
  const text = el.textContent?.trim();
  if (text) meta.click_target_text = text.slice(0, 120);
  return meta;
}

// src/utils/dedupeStorage.ts
var PREFIX = "probat:seen:";
var memorySet = /* @__PURE__ */ new Set();
function makeDedupeKey(experimentId, variantKey, instanceId, pageKey) {
  return `${PREFIX}${experimentId}:${variantKey}:${instanceId}:${pageKey}`;
}
function hasSeen(key) {
  if (memorySet.has(key)) return true;
  if (typeof window === "undefined") return false;
  try {
    return sessionStorage.getItem(key) === "1";
  } catch {
    return false;
  }
}
function markSeen(key) {
  memorySet.add(key);
  if (typeof window === "undefined") return;
  try {
    sessionStorage.setItem(key, "1");
  } catch {
  }
}
var INSTANCE_PREFIX = "probat:instance:";
function shortId() {
  const bytes = new Uint8Array(4);
  if (typeof crypto !== "undefined" && crypto.getRandomValues) {
    crypto.getRandomValues(bytes);
  } else {
    for (let i = 0; i < 4; i++) bytes[i] = Math.floor(Math.random() * 256);
  }
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}
function resolveStableId(storageKey) {
  if (typeof window !== "undefined") {
    try {
      const stored = sessionStorage.getItem(storageKey);
      if (stored) return stored;
    } catch {
    }
  }
  const id = `inst_${shortId()}`;
  if (typeof window !== "undefined") {
    try {
      sessionStorage.setItem(storageKey, id);
    } catch {
    }
  }
  return id;
}
var slotCounters = /* @__PURE__ */ new Map();
var resetScheduled = false;
function claimSlot(groupKey) {
  const idx = slotCounters.get(groupKey) ?? 0;
  slotCounters.set(groupKey, idx + 1);
  if (!resetScheduled) {
    resetScheduled = true;
    Promise.resolve().then(() => {
      slotCounters.clear();
      resetScheduled = false;
    });
  }
  return idx;
}
function useStableInstanceIdV18(experimentId) {
  const reactId = React3__default.default.useId();
  const ref = React3.useRef("");
  if (!ref.current) {
    const key = `${INSTANCE_PREFIX}${experimentId}:${getPageKey()}:${reactId}`;
    ref.current = resolveStableId(key);
  }
  return ref.current;
}
function useStableInstanceIdFallback(experimentId) {
  const slotRef = React3.useRef(-1);
  const ref = React3.useRef("");
  if (slotRef.current === -1) {
    slotRef.current = claimSlot(`${experimentId}:${getPageKey()}`);
  }
  if (!ref.current) {
    const key = `${INSTANCE_PREFIX}${experimentId}:${getPageKey()}:${slotRef.current}`;
    ref.current = resolveStableId(key);
  }
  return ref.current;
}
var useStableInstanceId = typeof React3__default.default.useId === "function" ? useStableInstanceIdV18 : useStableInstanceIdFallback;

// src/components/Experiment.tsx
var ASSIGNMENT_PREFIX = "probat:assignment:";
function readAssignment(id) {
  if (typeof window === "undefined") return null;
  try {
    const raw = localStorage.getItem(ASSIGNMENT_PREFIX + id);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    return parsed.variantKey ?? null;
  } catch {
    return null;
  }
}
function writeAssignment(id, variantKey) {
  if (typeof window === "undefined") return;
  try {
    const entry = { variantKey, ts: Date.now() };
    localStorage.setItem(ASSIGNMENT_PREFIX + id, JSON.stringify(entry));
  } catch {
  }
}
function Experiment({
  id,
  control,
  variants,
  track,
  componentInstanceId,
  fallback = "control",
  debug = false
}) {
  const { host, bootstrap } = useProbatContext();
  const autoInstanceId = useStableInstanceId(id);
  const instanceId = componentInstanceId ?? autoInstanceId;
  const trackImpression = track?.impression !== false;
  const trackClick = track?.primaryClick !== false;
  const impressionEvent = track?.impressionEventName ?? "$experiment_exposure";
  const clickEvent = track?.clickEventName ?? "$experiment_click";
  const [variantKey, setVariantKey] = React3.useState(() => {
    if (bootstrap[id]) return bootstrap[id];
    const cached = readAssignment(id);
    if (cached) return cached;
    return "control";
  });
  const [resolved, setResolved] = React3.useState(() => {
    return !!(bootstrap[id] || readAssignment(id));
  });
  React3.useEffect(() => {
    if (bootstrap[id] || readAssignment(id)) {
      const key = bootstrap[id] ?? readAssignment(id) ?? "control";
      setVariantKey(key);
      setResolved(true);
      return;
    }
    let cancelled = false;
    (async () => {
      try {
        const distinctId = getDistinctId();
        const key = await fetchDecision(host, id, distinctId);
        if (cancelled) return;
        if (key !== "control" && !(key in variants)) {
          if (debug) {
            console.warn(
              `[probat] Unknown variant "${key}" for experiment "${id}", falling back to control`
            );
          }
          setVariantKey("control");
        } else {
          setVariantKey(key);
          writeAssignment(id, key);
        }
      } catch (err) {
        if (cancelled) return;
        if (debug) {
          console.error(`[probat] fetchDecision failed for "${id}":`, err);
        }
        if (fallback === "suspend") throw err;
        setVariantKey("control");
      } finally {
        if (!cancelled) setResolved(true);
      }
    })();
    return () => {
      cancelled = true;
    };
  }, [id, host]);
  React3.useEffect(() => {
    if (debug && resolved) {
      console.log(`[probat] Experiment "${id}" \u2192 variant "${variantKey}"`, {
        instanceId,
        pageKey: getPageKey()
      });
    }
  }, [debug, id, variantKey, resolved, instanceId]);
  const eventProps = React3.useMemo(
    () => ({
      experiment_id: id,
      variant_key: variantKey,
      component_instance_id: instanceId
    }),
    [id, variantKey, instanceId]
  );
  const containerRef = React3.useRef(null);
  const impressionSent = React3.useRef(false);
  React3.useEffect(() => {
    if (!trackImpression || !resolved) return;
    impressionSent.current = false;
    const pageKey = getPageKey();
    const dedupeKey = makeDedupeKey(id, variantKey, instanceId, pageKey);
    if (hasSeen(dedupeKey)) {
      impressionSent.current = true;
      return;
    }
    const el = containerRef.current;
    if (!el) return;
    if (typeof IntersectionObserver === "undefined") {
      if (!impressionSent.current) {
        impressionSent.current = true;
        markSeen(dedupeKey);
        sendMetric(host, impressionEvent, eventProps);
        if (debug) console.log(`[probat] Impression sent (no IO) for "${id}"`);
      }
      return;
    }
    let timer = null;
    const observer = new IntersectionObserver(
      ([entry]) => {
        if (!entry || impressionSent.current) return;
        if (entry.isIntersecting) {
          timer = setTimeout(() => {
            if (impressionSent.current) return;
            impressionSent.current = true;
            markSeen(dedupeKey);
            sendMetric(host, impressionEvent, eventProps);
            if (debug) console.log(`[probat] Impression sent for "${id}"`);
            observer.disconnect();
          }, 250);
        } else if (timer) {
          clearTimeout(timer);
          timer = null;
        }
      },
      { threshold: 0.5 }
    );
    observer.observe(el);
    return () => {
      observer.disconnect();
      if (timer) clearTimeout(timer);
    };
  }, [
    trackImpression,
    resolved,
    id,
    variantKey,
    instanceId,
    host,
    impressionEvent,
    eventProps,
    debug
  ]);
  const handleClick = React3.useCallback(
    (e) => {
      if (!trackClick) return;
      const meta = extractClickMeta(e.target);
      if (!meta) return;
      sendMetric(host, clickEvent, {
        ...eventProps,
        ...meta
      });
      if (debug) {
        console.log(`[probat] Click tracked for "${id}"`, meta);
      }
    },
    [trackClick, host, clickEvent, eventProps, id, debug]
  );
  const content = variantKey === "control" || !(variantKey in variants) ? control : variants[variantKey];
  return /* @__PURE__ */ React3__default.default.createElement(
    "div",
    {
      ref: containerRef,
      onClick: handleClick,
      "data-probat-experiment": id,
      "data-probat-variant": variantKey,
      style: {
        display: "block",
        margin: 0,
        padding: 0,
        opacity: resolved ? 1 : 0,
        transition: resolved ? "opacity 0.15s ease-in" : "none"
      }
    },
    content
  );
}
function useProbatMetrics() {
  const { host } = useProbatContext();
  const capture = React3.useCallback(
    (event, properties = {}) => {
      sendMetric(host, event, properties);
    },
    [host]
  );
  return { capture };
}

exports.Experiment = Experiment;
exports.ProbatProviderClient = ProbatProviderClient;
exports.fetchDecision = fetchDecision;
exports.sendMetric = sendMetric;
exports.useProbatMetrics = useProbatMetrics;
