// @probat/react/src/utils/api.ts
import { detectEnvironment } from "./environment";
import { buildEventContext } from "./eventContext";

// ── Types ──────────────────────────────────────────────────────────────────

export interface DecisionResponse {
    variant_key: string;
}

export interface MetricPayload {
    event: string;
    environment: "dev" | "prod";
    properties: Record<string, unknown>;
}

// ── Assignment fetching ────────────────────────────────────────────────────

const pendingDecisions = new Map<string, Promise<string>>();

/**
 * Fetch the variant assignment for an experiment.
 * Returns the variant key string (e.g. "control", "ai_v1").
 * Deduplicates concurrent calls for the same experiment.
 */
export async function fetchDecision(
    host: string,
    experimentId: string,
    distinctId: string
): Promise<string> {
    const existing = pendingDecisions.get(experimentId);
    if (existing) return existing;

    const promise = (async () => {
        try {
            const url = `${host.replace(/\/$/, "")}/experiment/decide`;
            const res = await fetch(url, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    Accept: "application/json",
                },
                credentials: "include",
                body: JSON.stringify({
                    experiment_id: experimentId,
                    distinct_id: distinctId,
                }),
            });
            if (!res.ok) throw new Error(`HTTP ${res.status}`);
            const data = (await res.json()) as DecisionResponse;
            return data.variant_key || "control";
        } finally {
            pendingDecisions.delete(experimentId);
        }
    })();

    pendingDecisions.set(experimentId, promise);
    return promise;
}

// ── Metric sending ─────────────────────────────────────────────────────────

/**
 * Fire-and-forget metric send. Never throws.
 */
export function sendMetric(
    host: string,
    event: string,
    properties: Record<string, unknown>
): void {
    if (typeof window === "undefined") return;

    const ctx = buildEventContext();
    const payload: MetricPayload = {
        event,
        environment: detectEnvironment(),
        properties: {
            ...ctx,
            source: "react-sdk",
            captured_at: new Date().toISOString(),
            ...properties,
        },
    };

    try {
        const url = `${host.replace(/\/$/, "")}/experiment/metrics`;
        fetch(url, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            credentials: "include",
            body: JSON.stringify(payload),
        }).catch(() => {});
    } catch {
        // silently drop
    }
}

// ── Click metadata extraction ──────────────────────────────────────────────

export interface ClickMeta {
    click_target_tag: string;
    click_target_text?: string;
    click_target_id?: string;
    click_is_primary: boolean;
}

/**
 * Given a click event inside an experiment boundary, extract metadata.
 * Prioritizes elements with data-probat-click="primary",
 * then falls back to button/a/role=button.
 */
export function extractClickMeta(target: EventTarget | null): ClickMeta | null {
    if (!target || !(target instanceof HTMLElement)) return null;

    // Priority 1: explicit primary marker
    const primary = target.closest('[data-probat-click="primary"]');
    if (primary) return buildMeta(primary as HTMLElement, true);

    // Priority 2: interactive elements
    const interactive = target.closest('button, a, [role="button"]');
    if (interactive) return buildMeta(interactive as HTMLElement, false);

    return null;
}

function buildMeta(el: HTMLElement, isPrimary: boolean): ClickMeta {
    const meta: ClickMeta = {
        click_target_tag: el.tagName,
        click_is_primary: isPrimary,
    };
    if (el.id) meta.click_target_id = el.id;
    const text = el.textContent?.trim();
    if (text) meta.click_target_text = text.slice(0, 120);
    return meta;
}
