// @probat/react/src/utils/dedupeStorage.ts
/**
 * Dedupe storage for experiment exposures.
 * Uses sessionStorage + in-memory Set fallback.
 * Key format: probat:seen:{id}:{variantKey}:{instanceId}:{pageKey}
 */

const PREFIX = "probat:seen:";
const memorySet = new Set<string>();

export function makeDedupeKey(
    experimentId: string,
    variantKey: string,
    instanceId: string,
    pageKey: string
): string {
    return `${PREFIX}${experimentId}:${variantKey}:${instanceId}:${pageKey}`;
}

export function hasSeen(key: string): boolean {
    if (memorySet.has(key)) return true;
    if (typeof window === "undefined") return false;
    try {
        return sessionStorage.getItem(key) === "1";
    } catch {
        return false;
    }
}

export function markSeen(key: string): void {
    memorySet.add(key);
    if (typeof window === "undefined") return;
    try {
        sessionStorage.setItem(key, "1");
    } catch {}
}

/** Reset all dedupe state — useful for testing. */
export function resetDedupe(): void {
    memorySet.clear();
}
