// @probat/react/src/utils/eventContext.ts
/**
 * Event context helpers: distinct_id, session_id, page info.
 * All browser-safe — no-ops when window is unavailable.
 */

const DISTINCT_ID_KEY = "probat:distinct_id";
const SESSION_ID_KEY = "probat:session_id";

let cachedDistinctId: string | null = null;
let cachedSessionId: string | null = null;

function generateId(): string {
    // crypto.randomUUID where available, else fallback
    if (typeof crypto !== "undefined" && crypto.randomUUID) {
        return crypto.randomUUID();
    }
    // fallback: random hex
    const bytes = new Uint8Array(16);
    if (typeof crypto !== "undefined" && crypto.getRandomValues) {
        crypto.getRandomValues(bytes);
    } else {
        for (let i = 0; i < 16; i++) bytes[i] = Math.floor(Math.random() * 256);
    }
    return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

export function getDistinctId(): string {
    if (cachedDistinctId) return cachedDistinctId;
    if (typeof window === "undefined") return "server";
    try {
        const stored = localStorage.getItem(DISTINCT_ID_KEY);
        if (stored) {
            cachedDistinctId = stored;
            return stored;
        }
    } catch {}
    const id = `anon_${generateId()}`;
    cachedDistinctId = id;
    try {
        localStorage.setItem(DISTINCT_ID_KEY, id);
    } catch {}
    return id;
}

export function getSessionId(): string {
    if (cachedSessionId) return cachedSessionId;
    if (typeof window === "undefined") return "server";
    try {
        const stored = sessionStorage.getItem(SESSION_ID_KEY);
        if (stored) {
            cachedSessionId = stored;
            return stored;
        }
    } catch {}
    const id = `sess_${generateId()}`;
    cachedSessionId = id;
    try {
        sessionStorage.setItem(SESSION_ID_KEY, id);
    } catch {}
    return id;
}

export function getPageKey(): string {
    if (typeof window === "undefined") return "";
    return window.location.pathname + window.location.search;
}

export function getPageUrl(): string {
    if (typeof window === "undefined") return "";
    return window.location.href;
}

export function getReferrer(): string {
    if (typeof document === "undefined") return "";
    return document.referrer;
}

export interface EventContext {
    distinct_id: string;
    session_id: string;
    $page_url: string;
    $pathname: string;
    $referrer: string;
}

export function buildEventContext(): EventContext {
    return {
        distinct_id: getDistinctId(),
        session_id: getSessionId(),
        $page_url: getPageUrl(),
        $pathname: typeof window !== "undefined" ? window.location.pathname : "",
        $referrer: getReferrer(),
    };
}
