// @probat/react/src/utils/stableInstanceId.ts
/**
 * Stable auto-generated instance IDs for <Experiment />.
 *
 * Problem: a naïve useRef + module counter gives a different ID on every mount,
 * so StrictMode double-mount or unmount/remount changes the dedupe key.
 *
 * Solution:
 *   1. React 18+  → useId() is stable per fiber position.
 *   2. Fallback   → sessionStorage-backed slot counter per (experimentId, pageKey).
 *   3. Both paths persist a mapping in sessionStorage:
 *        probat:instance:{experimentId}:{pageKey}:{positionKey} → stableId
 *      so the same position resolves to the same ID across mounts.
 */

import React, { useRef } from "react";
import { getPageKey } from "./eventContext";

const INSTANCE_PREFIX = "probat:instance:";

// ── Helpers ────────────────────────────────────────────────────────────────

function shortId(): string {
    const bytes = new Uint8Array(4);
    if (typeof crypto !== "undefined" && crypto.getRandomValues) {
        crypto.getRandomValues(bytes);
    } else {
        for (let i = 0; i < 4; i++) bytes[i] = Math.floor(Math.random() * 256);
    }
    return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

/**
 * Look up or create a stable instance ID in sessionStorage.
 */
function resolveStableId(storageKey: string): string {
    if (typeof window !== "undefined") {
        try {
            const stored = sessionStorage.getItem(storageKey);
            if (stored) return stored;
        } catch {}
    }
    const id = `inst_${shortId()}`;
    if (typeof window !== "undefined") {
        try {
            sessionStorage.setItem(storageKey, id);
        } catch {}
    }
    return id;
}

// ── Fallback: render-wave slot counter ─────────────────────────────────────
// Each synchronous render batch claims sequential slots per (experimentId,
// pageKey). A microtask resets the counters so the next batch starts at 0,
// giving the same component position the same slot across mounts.

const slotCounters = new Map<string, number>();
let resetScheduled = false;

function claimSlot(groupKey: string): number {
    const idx = slotCounters.get(groupKey) ?? 0;
    slotCounters.set(groupKey, idx + 1);
    if (!resetScheduled) {
        resetScheduled = true;
        Promise.resolve().then(() => {
            slotCounters.clear();
            resetScheduled = false;
        });
    }
    return idx;
}

// ── Hook: React 18+ path (useId available) ─────────────────────────────────

function useStableInstanceIdV18(experimentId: string): string {
    const reactId = (React as any).useId() as string;
    const ref = useRef("");
    if (!ref.current) {
        const key = `${INSTANCE_PREFIX}${experimentId}:${getPageKey()}:${reactId}`;
        ref.current = resolveStableId(key);
    }
    return ref.current;
}

// ── Hook: fallback path (no useId) ─────────────────────────────────────────

function useStableInstanceIdFallback(experimentId: string): string {
    const slotRef = useRef(-1);
    const ref = useRef("");
    if (slotRef.current === -1) {
        slotRef.current = claimSlot(`${experimentId}:${getPageKey()}`);
    }
    if (!ref.current) {
        const key = `${INSTANCE_PREFIX}${experimentId}:${getPageKey()}:${slotRef.current}`;
        ref.current = resolveStableId(key);
    }
    return ref.current;
}

// ── Exported hook ──────────────────────────────────────────────────────────
// Selection is a module-level constant so the hook-call count never changes
// between renders — safe for the rules of hooks.

export const useStableInstanceId: (experimentId: string) => string =
    typeof (React as any).useId === "function"
        ? useStableInstanceIdV18
        : useStableInstanceIdFallback;

// ── Test utility ───────────────────────────────────────────────────────────

export function resetInstanceIdState(): void {
    slotCounters.clear();
    resetScheduled = false;
}
